devdoc/channel_2.md

```diff
@@ -0,0 +1,72 @@
+# channel_2 requirements:
+
+## Overview
+
+The current implementation of `channel` has certain pitfalls. `channel_2` aims to address these pitfalls by making the following changes:
+
+1. Introduce a `channel_2_reset` API: This API will allow the user to reset the `channel_2` to a clean state. It will cause all pending operations to be abandoned.
+
+2. Remove `ASYNC_OP`: The current implementation of `channel` provides the user with an `ASYNC_OP` when an operation (`channel_push` or `channel_pull`) is initiated. The user can call `async_op_cancel` to do a best-effort cancellation of the operation. There are 2 issues with this:
+    - Complicates object lifetimes: The lifetime of the operation must be linked to the `ASYNC_OP`and the `channel`, since each can be independently destroyed by the user.
```

Comment:
>each can be independently destroyed by the use [](http://example.com/codeflow?start=123&length=46)

I _really_ don't like the word "destroyed" here because channel and async_op are both THANDLEs and THANDLEs can't be "destroyed" by the user. They can be decref'ed , but the can't be destroyed.

This is why we are adding open & close. The user can close the channel, and they can decref the channel, but they cannot destroy the channel.

It may feel like I'm arguing semantics, but "close", "destroy", and "dispose" are 3 very different concepts, and we're in this mess because the existing design of channel combines them all together.